design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,294 @@
+service execution result may contain the return value of a service or an exception it has thrown. The promise that invoker returns may
+be used in both synchronous and asynchronous ways to be able to work with both types of services. More details about the promise
+[here](../promises.md).
+ 
+Proxy DTOs will look like this:
+```php
+namespace Magento\ServiceBProxy;
+
+class BDataProxy implements BDataInterface
+{
+    private $id;
+    
+    private $name;
+    
+    public function __construct(string $name, ?string $id) {
+        $this->name = $name;
+        $this->id = $id;
+    }
+    
+    public function getId(): ?string
+    {
+        return $this->id;
+    }
+    
+    public function setId(string $id): void
+    {
+        $this->>id = $id;
+    }
+    
+    public function getName(): string
+    {
+        return $this->name;
+    }
+    
+    public function setName(string $name): void
+    {
+        $this->name = $name;
+    }
+}
+```
+Usually DTOs implementations are models in real modules but we need a class to be initialized when client code communicates with a
+remote service and for that simple DTOs will suffice.
+ 
+#### The invoker
+The invoker will be a part of Magento\Framework and will be present in every node installation. Invoker's job is to find a remote node's
+address based on given class name and method, send the request, receive response and convert it to either requested service method return type
+or an exception if it was thrown by the remote service. It returns a promise to support asynchronous services and because
+the requests to remote services will be asynchronous (depends on implementation). More details on the invokers API and SPI
+[here](invoker.md).
+ 
+#### Gateway
+For actualy delivering remote service requests RESTful API will be used - it's an established way to execute requested methods and to
+serialize arguments and return values. For sending HTTP requests [Guzzle](https://github.com/guzzle/guzzle) library will be used - it's
+a stable and well supported library which is capable of sending concurrent requests and allows wide customization.
+Guzzle will be configured to use cUrl (cUrl multi) as a backend. When asynchronous requests are sent via Guzzle it returns a promise which can easily
+be converted to Magento's promise.
+#### Lookup
+The information on how to reach services will be provided in env.php like this:
+```php
+return [
+    'service_base_urls' => [
+        'Magento\\Catalog' => 'http://catalog.mydomain.com/',
+        'Magento\\Checkout' => 'https://checkout.mydomain.com/',
+        'Magento\\Customer' => 'http://192.168.1.1/',
+    ],
+];
+```
+The invoker will look up the information based on provided services' class names.
+ 
+#### Changes to RESTful Web API mechanism
+We need to able so send and receive more information when using RESTful web API for inter-service communication than for regular
```

Comment:
What do you mean by regular endpoints, public APIs? Public API will be defined on the BFF. Services define private APIs. Only other services and BFF can use APIs of the services. Based on this we have clear distinction.

Reply:
I would suggest to call it public, because it could be confusing.